File: src/main/java/com/zoarial/iot/threads/tcp/TCPThread.java
package com.zoarial.iot.threads.tcp;

import com.zoarial.*;
import com.zoarial.iot.action.dao.IoTActionDAO;
import com.zoarial.iot.dto.ZoarialDTO;
import com.zoarial.iot.node.dao.IoTNodeDAO;
import com.zoarial.iot.node.model.IoTNode;
import com.zoarial.iot.action.model.IoTAction;
import com.zoarial.iot.server.ServerServer;
import com.zoarial.iot.network.IoTPacketSectionList;
import com.zoarial.iot.network.IoTSession;
import com.zoarial.iot.threads.tcp.networkModel.TCPStart;
import me.zoarial.networkArbiter.ZoarialNetworkArbiter;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.EOFException;
import java.io.IOException;
import java.net.SocketException;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

// TODO: redo the logic here. It may not close properly in all cases.
public class TCPThread extends PrintBaseClass implements Runnable {

    private final ServerServer _server;
    private static final AtomicInteger idNumber = new AtomicInteger(0);
    private static final List<Thread> threads = Collections.synchronizedList(new ArrayList<>(8));
    private final byte _version = (byte)0;
    HashMap<Integer, IoTSession> sessions = new HashMap<>();

    private final SocketHelper _inSocket;
    private final IoTNodeDAO ioTNodeDAO = new IoTNodeDAO();
    private final IoTActionDAO ioTActionDAO = new IoTActionDAO();

    private final ZoarialNetworkArbiter networkArbiter = ZoarialNetworkArbiter.getInstance();

    public TCPThread(ServerServer server, SocketHelper inSocket) {
        super("TCPThread" + idNumber.getAndIncrement());
        _inSocket = inSocket;
        _server = server;
    }

    public void run() {
        // TODO: have some way to add and remove the socket to the list
        // We could then close the socket and have it flush before the ServerSocket is closed
        // UPDATE: we should use shutdown to shutdown the input, then clean up, then close.
        threads.add(Thread.currentThread());
        println("Starting connection with: " + _inSocket.inSocket.getRemoteSocketAddress());
        try { // This might be unnecessary, but better safe than sorry
            runLogic();
        } finally {
            // Make sure even if something goes wrong, that these things happen.
            threads.remove(Thread.currentThread());
            cleanup();
            println("Stopping.");
        }
    }

    // TODO: redo this to use a circular buffer for the input
    // may not need circular buffer. Maybe assume all data coming in is correct and if not, then the socket will eventually close.
    // Use the raw stream and data stream
    private void runLogic() {
        // If we leave this try/catch block, then the socket is closed.
        while (!_server.isClosed() && !_inSocket.inSocket.isClosed()) {
            String request;
            int otp;
            byte securityLevel;
            try {
                securityLevel = 0;

                TCPStart startInfo = networkArbiter.receiveObject(TCPStart.class, _inSocket.inSocket);
                if (startInfo.getVersion() != _version) {
                    println("Not the correct version.");
                    return;
                }

                int sessionID = startInfo.getSessionId();
                if (sessions.containsKey(sessionID)) {
                    // Continue working with session
                    println("Continuing to work with session: " + sessionID);
                } else {
                    IoTSession session = new IoTSession(sessionID);
                    // New session
                    request = startInfo.getRequest();
                    println("Working with new session: " + sessionID);
                    println("Request: " + request);
                    switch (request) {
                        case "action" -> {
                            var actionData = networkArbiter.receiveObject(ZoarialDTO.V1.Request.Action.class, _inSocket.inSocket);

                            JSONObject jsonObject = new JSONObject(actionData.getData());

                            if(actionData.getOtp() != null) {
                                securityLevel = 1;
                            }

                            JSONArray args = jsonObject.getJSONArray("args");
                            UUID uuid = actionData.getUuid();
                            // Check for the action
                            var optAction = ioTActionDAO.findActionByUUID(uuid);
                            if(optAction.isPresent()) {
                                var action = optAction.get();
                                if(!action.getNode().equals(_server.getSelfNode())) {
                                    respondToSession(session, _server.runRemoteAction(action, jsonObject));
                                } else if(action.isLocal() && !_inSocket.isLocal()) {
                                    respondToSession(session, "You don't have access to this. (Local).");
                                    // Check assess
                                } else if(action.isEncrypted() && !(_inSocket.isEncrypted() || _inSocket.isLocal())) {
                                    respondToSession(session, "You don't have access to this. (Encrypted).");
                                    // You have access, start doing stuff
                                } else {

                                    int numOfArguments = action.getArguments();
                                    if (numOfArguments == 0) {
                                        request = action.execute();
                                    } else {
                                        request = action.execute(args);
                                    }
                                    respondToSession(session, request);
                                }
                            } else {
                                respondToSession(session, "No action with UUID: " + uuid);
                            }
                        }
                        case "updateActionSecurityLevel" -> {
                            var updateInfo = networkArbiter.receiveObject(ZoarialDTO.V1.Request.UpdateActionSecurityLevel.class, _inSocket.inSocket);
                            UUID uuid = updateInfo.getUuid();
                            println("Updating security level of " + uuid + " to: " + updateInfo.getLevel());

                            var optAction = ioTActionDAO.findActionByUUID(uuid);
                            if(optAction.isEmpty()) {
                                respondToSession(session, "No action with UUID: " + uuid);
                                break;
                            }

                            var action = optAction.get();
                            if(!action.getNode().equals(_server.getSelfNode())) {
                                respondToSession(session, "I can't update other node's actions.");
                                break;
                            }

                            action.setSecurityLevel(updateInfo.getLevel());
                            ioTActionDAO.update(action);
                            respondToSession(session, "Success.");
                        }
                        case "updateActionEncrypt" -> {
                            var updateInfo = networkArbiter.receiveObject(ZoarialDTO.V1.Request.UpdateActionEncrypt.class, _inSocket.inSocket);
                            UUID uuid = updateInfo.getUuid();
                            println("Updating encrypt toggle of " + uuid + " to :" + updateInfo.isEncrypt());
                            var optAction = ioTActionDAO.findActionByUUID(uuid);
                            if(optAction.isEmpty()) {
                                respondToSession(session, "No action with UUID: " + uuid);
                                break;
                            }
                            var action = optAction.get();
                            if(!action.getNode().equals(_server.getSelfNode())) {
                                respondToSession(session, "I can't update other node's actions.");
                                break;
                            }
                            action.setEncrypted(updateInfo.isEncrypt());
                            ioTActionDAO.update(action);
                            respondToSession(session, "Success.");

                        }
                        case "updateActionLocal" -> {
                            var updateInfo = networkArbiter.receiveObject(ZoarialDTO.V1.Request.UpdateActionLocal.class, _inSocket.inSocket);
                            UUID uuid = updateInfo.getUuid();
                            println("Updating local toggle of " + uuid + " to: " + updateInfo.isLocal());

                            var optAction = ioTActionDAO.findActionByUUID(uuid);
                            if(optAction.isEmpty()) {
                                respondToSession(session, "No action with UUID: " + uuid);
                                break;
                            }

                            var action = optAction.get();
                            if(!action.getNode().equals(_server.getSelfNode())) {
                                respondToSession(session, "I can't update other node's actions.");
                                break;
                            }

                            action.setLocal(updateInfo.isLocal());
                            ioTActionDAO.update(action);
                            respondToSession(session, "Success.");
                        }
                        case "updateActionDescription" -> {
                            var updateInfo = networkArbiter.receiveObject(ZoarialDTO.V1.Request.UpdateActionDescription.class, _inSocket.inSocket);
                            UUID uuid = updateInfo.getUuid();
                            println("Updating local toggle of " + uuid + " to: " + updateInfo.getDescription());

                            var optAction = ioTActionDAO.findActionByUUID(uuid);
                            if(optAction.isEmpty()) {
                                respondToSession(session, "No action with UUID: " + uuid);
                                break;
                            }

                            var action = optAction.get();
                            if(!action.getNode().equals(_server.getSelfNode())) {
                                respondToSession(session, "I can't update other node's actions.");
                                break;
                            }

                            action.setDescription(updateInfo.getDescription());
                            ioTActionDAO.update(action);
                            respondToSession(session, "Success.");

                        }
                        case "info" -> {
                            request = _inSocket.readString();
                            println("Request info: " + request);
                            switch (request) {
                                case "action" -> {
                                    UUID actionUUID = _inSocket.readUUID();
                                    println("Responding info about action: " + actionUUID);
                                    IoTAction action = ioTActionDAO.getActionByUUID(actionUUID);
                                    IoTPacketSectionList sectionList = new IoTPacketSectionList();
                                    if(action == null) {
                                        sectionList.add((byte)0);
                                        sectionList.add("An action with that UUID does not exist.");
                                        println("Action does not exist.");
                                    } else {
                                        println("Responding with action info.");
                                        sectionList.add((byte)1);
                                        sectionList.add(action.toJson().toString());
                                    }
                                    respondToSession(new IoTSession(sessionID, IoTSession.IoTSessionType.INFO), sectionList);
                                }
                                case "actions" -> respondActionList(new IoTSession(sessionID, IoTSession.IoTSessionType.INFO));
                                case "general" -> {
                                    IoTPacketSectionList sectionList = new IoTPacketSectionList();
                                    sectionList.add(_server.getServerInfo().hostname);
                                    sectionList.add(_server.getServerInfo().uuid);
                                    sectionList.add(_server.getServerInfo().isHeadCapable);
                                    sectionList.add(_server.getServerInfo().isVolatile);
                                    respondToSession(new IoTSession(sessionID, IoTSession.IoTSessionType.INFO), sectionList);
                                }
                                case "nodes" -> {
                                    respondNodeList(new IoTSession(sessionID, IoTSession.IoTSessionType.INFO));
                                }
                                default -> respondToSession(new IoTSession(sessionID, IoTSession.IoTSessionType.INFO), "Unknown option to get: " + request);
                            }
                        }
                        default -> respondToSession(new IoTSession(sessionID, IoTSession.IoTSessionType.OTHER), "Invalid request: " + request);
                    }
                }

                // We have to flush otherwise, out data may never be sent.
                // TODO: fixing the shutdown would prevent the need for this
                _inSocket.out.flush();
            } catch (EOFException e) {
                println("End of file: should be closed!");
                return; // Cleanup is done after function returns
            } catch (SocketException ex) {
                println("Interrupted, server must be closing.");
                return; // Cleanup is done after function returns
            } catch (IOException ex) {
                ex.printStackTrace();
                return; // Cleanup is done after function returns
            } catch (JSONException ex) {
                ex.printStackTrace();
                println("Something is wrong with the json format.");
                return;
            }
        }
    }

    private void respondToSession(IoTSession session,  IoTPacketSectionList sectionList) {
        try {
            _inSocket.out.writeBytes("ZIoT");
            _inSocket.out.writeInt(session.getSessionID());
            _inSocket.out.write(sectionList.getNetworkResponse());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void respondToSession(IoTSession session, String str) {
        println("Responding: " + str);
        IoTPacketSectionList sectionList = new IoTPacketSectionList();
        sectionList.add("ZIoT");
        sectionList.add(session.getSessionID());
        sectionList.add(str);
        try {
            _inSocket.out.write(sectionList.getNetworkResponse());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void respondActionList(IoTSession session) {
        var actionList = ioTActionDAO.getEnabledActions();
        IoTPacketSectionList sectionList = new IoTPacketSectionList(actionList.size() * 4 + 3);

        sectionList.add("ZIoT");
        sectionList.add(session.getSessionID());
        sectionList.add(actionList.size());

        for(IoTAction action : actionList) {
            sectionList.add(action.getUuid());
            sectionList.add(action.getNode().getUuid());
            sectionList.add(action.getName());
            //sectionList.add(action.getReturnType())
            sectionList.add(action.getSecurityLevel());
            sectionList.add(action.getArguments());
            sectionList.add(action.isEncrypted());
            sectionList.add(action.isLocal());
        }

        try {
            _inSocket.out.write(sectionList.getNetworkResponse());
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private void respondNodeList(IoTSession session) {
        IoTNodeDAO ioTNodeDAO = new IoTNodeDAO();
        var nodeList = ioTNodeDAO.getAllNodes();
        IoTPacketSectionList sectionList = new IoTPacketSectionList(nodeList.size() * 4 + 3);


        sectionList.add("ZIoT");
        sectionList.add(session.getSessionID());
        sectionList.add(nodeList.size());

        for(IoTNode node : nodeList) {
            sectionList.add(node.getUuid());
            sectionList.add(node.getHostname());
            sectionList.add(node.getNodeType());
            sectionList.add(node.getLastHeardFrom());
        }

        try {
            _inSocket.out.write(sectionList.getNetworkResponse());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    static List<Thread> getThreads() {
        return threads;
    }

    private void cleanup() {
        try {
            if(!_inSocket.isClosed()) {
                _inSocket.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: src/main/java/com/zoarial/iot/threads/tcp/TCPThread.kt
package com.zoarial.iot.threads.tcp

import com.zoarial.PrintBaseClass
import com.zoarial.iot.action.dao.IoTActionDAO
import com.zoarial.iot.dto.ZoarialDTO
import com.zoarial.iot.dto.ZoarialDTO.V1.Request.*
import com.zoarial.iot.network.IoTPacketSectionList
import com.zoarial.iot.network.IoTSession
import com.zoarial.iot.network.IoTSession.IoTSessionType
import com.zoarial.iot.node.dao.IoTNodeDAO
import com.zoarial.iot.server.ServerServer
import com.zoarial.iot.threads.tcp.networkModel.TCPStart
import me.zoarial.networkArbiter.ZoarialNetworkArbiter
import org.json.JSONException
import org.json.JSONObject
import java.io.EOFException
import java.io.IOException
import java.net.SocketException
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

// TODO: redo the logic here. It may not close properly in all cases.
class TCPThread(private val _server: ServerServer, private val _inSocket: SocketHelper) : PrintBaseClass("TCPThread" + idNumber.getAndIncrement()), Runnable {
    private val _version = 0.toByte()
    var sessions = HashMap<Int, IoTSession>()
    private val ioTNodeDAO = IoTNodeDAO()
    private val ioTActionDAO = IoTActionDAO()
    private val networkArbiter: ZoarialNetworkArbiter = ZoarialNetworkArbiter
    override fun run() {
        // TODO: have some way to add and remove the socket to the list
        // We could then close the socket and have it flush before the ServerSocket is closed
        // UPDATE: we should use shutdown to shutdown the input, then clean up, then close.
        threads.add(Thread.currentThread())
        println("Starting connection with: " + _inSocket.inSocket.remoteSocketAddress)
        try { // This might be unnecessary, but better safe than sorry
            runLogic()
        } finally {
            // Make sure even if something goes wrong, that these things happen.
            threads.remove(Thread.currentThread())
            cleanup()
            println("Stopping.")
        }
    }

    // TODO: redo this to use a circular buffer for the input
    // may not need circular buffer. Maybe assume all data coming in is correct and if not, then the socket will eventually close.
    // Use the raw stream and data stream
    private fun runLogic() {
        // If we leave this try/catch block, then the socket is closed.
        while (!_server.isClosed && !_inSocket.inSocket.isClosed) {
            var request: String?
            var otp: Int
            var securityLevel: Byte
            try {
                securityLevel = 0
                val startInfo = networkArbiter.receiveObject(TCPStart::class.java, _inSocket.inSocket)
                if (startInfo.version != _version) {
                    println("Not the correct version.")
                    return
                }
                val sessionID = startInfo.sessionId
                if (sessions.containsKey(sessionID)) {
                    // Continue working with session
                    println("Continuing to work with session: $sessionID")
                } else {
                    val session = IoTSession(sessionID)
                    // New session
                    request = startInfo.request
                    println("Working with new session: $sessionID")
                    println("Request: $request")
                    when (request) {
                        "action" -> {
                            val actionData = networkArbiter.receiveObject(ZoarialDTO.V1.Request.Action::class.java, _inSocket.inSocket)
                            val jsonObject = JSONObject(actionData.data)
                            if (actionData.otp != null) {
                                securityLevel = 1
                            }
                            val args = jsonObject.getJSONArray("args")
                            val uuid = actionData.uuid
                            // Check for the action
                            val optAction = ioTActionDAO.findActionByUUID(uuid!!)
                            if (optAction.isPresent) {
                                val action = optAction.get()
                                if (action.node != _server.selfNode) {
                                    respondToSession(session, _server.runRemoteAction(action, jsonObject))
                                } else if (action.local!! && !_inSocket.isLocal) {
                                    respondToSession(session, "You don't have access to this. (Local).")
                                    // Check assess
                                } else if (action.encrypted!! && !(_inSocket.isEncrypted || _inSocket.isLocal)) {
                                    respondToSession(session, "You don't have access to this. (Encrypted).")
                                    // You have access, start doing stuff
                                } else {
                                    val numOfArguments = action.arguments!!.toInt()
                                    request = if (numOfArguments == 0) {
                                        action.execute()
                                    } else {
                                        action.execute(args)
                                    }
                                    respondToSession(session, request)
                                }
                            } else {
                                respondToSession(session, "No action with UUID: $uuid")
                            }
                        }

                        "updateActionSecurityLevel" -> {
                            val updateInfo = networkArbiter.receiveObject(UpdateActionSecurityLevel::class.java, _inSocket.inSocket)
                            val uuid = updateInfo.uuid
                            println("Updating security level of " + uuid + " to: " + updateInfo.level)
                            val optAction = ioTActionDAO.findActionByUUID(uuid!!)
                            if (optAction.isEmpty) {
                                respondToSession(session, "No action with UUID: $uuid")
                                break
                            }
                            val action = optAction.get()
                            if (action.node != _server.selfNode) {
                                respondToSession(session, "I can't update other node's actions.")
                                break
                            }
                            action.securityLevel = updateInfo.level
                            ioTActionDAO.update(action)
                            respondToSession(session, "Success.")
                        }

                        "updateActionEncrypt" -> {
                            val updateInfo = networkArbiter.receiveObject(UpdateActionEncrypt::class.java, _inSocket.inSocket)
                            val uuid = updateInfo.uuid
                            println("Updating encrypt toggle of " + uuid + " to :" + updateInfo.encrypt)
                            val optAction = ioTActionDAO.findActionByUUID(uuid!!)
                            if (optAction.isEmpty) {
                                respondToSession(session, "No action with UUID: $uuid")
                                break
                            }
                            val action = optAction.get()
                            if (action.node != _server.selfNode) {
                                respondToSession(session, "I can't update other node's actions.")
                                break
                            }
                            action.encrypted = updateInfo.encrypt
                            ioTActionDAO.update(action)
                            respondToSession(session, "Success.")
                        }

                        "updateActionLocal" -> {
                            val updateInfo = networkArbiter.receiveObject(UpdateActionLocal::class.java, _inSocket.inSocket)
                            val uuid = updateInfo.uuid
                            println("Updating local toggle of " + uuid + " to: " + updateInfo.local)
                            val optAction = ioTActionDAO.findActionByUUID(uuid!!)
                            if (optAction.isEmpty) {
                                respondToSession(session, "No action with UUID: $uuid")
                                break
                            }
                            val action = optAction.get()
                            if (action.node != _server.selfNode) {
                                respondToSession(session, "I can't update other node's actions.")
                                break
                            }
                            action.local = updateInfo.local
                            ioTActionDAO.update(action)
                            respondToSession(session, "Success.")
                        }

                        "updateActionDescription" -> {
                            val updateInfo = networkArbiter.receiveObject(UpdateActionDescription::class.java, _inSocket.inSocket)
                            val uuid = updateInfo.uuid
                            println("Updating local toggle of " + uuid + " to: " + updateInfo.description)
                            val optAction = ioTActionDAO.findActionByUUID(uuid!!)
                            if (optAction.isEmpty) {
                                respondToSession(session, "No action with UUID: $uuid")
                                break
                            }
                            val action = optAction.get()
                            if (action.node != _server.selfNode) {
                                respondToSession(session, "I can't update other node's actions.")
                                break
                            }
                            action.description = updateInfo.description
                            ioTActionDAO.update(action)
                            respondToSession(session, "Success.")
                        }

                        "info" -> {
                            request = _inSocket.readString()
                            println("Request info: $request")
                            when (request) {
                                "action" -> {
                                    val actionUUID = _inSocket.readUUID()
                                    println("Responding info about action: $actionUUID")
                                    val actionOptional = ioTActionDAO.findActionByUUID(actionUUID)
                                    val sectionList = IoTPacketSectionList()
                                    if (actionOptional.isEmpty) {
                                        sectionList.add(0.toByte())
                                        sectionList.add("An action with that UUID does not exist.")
                                        println("Action does not exist.")
                                    } else {
                                        println("Responding with action info.")
                                        sectionList.add(1.toByte())
                                        sectionList.add(actionOptional.get().toJson().toString())
                                    }
                                    respondToSession(IoTSession(sessionID, IoTSessionType.INFO), sectionList)
                                }

                                "actions" -> respondActionList(IoTSession(sessionID, IoTSessionType.INFO))
                                "general" -> {
                                    val sectionList = IoTPacketSectionList()
                                    sectionList.add(_server.serverInfo.hostname)
                                    sectionList.add(_server.serverInfo.uuid)
                                    sectionList.add(_server.serverInfo.isHeadCapable)
                                    sectionList.add(_server.serverInfo.isVolatile)
                                    respondToSession(IoTSession(sessionID, IoTSessionType.INFO), sectionList)
                                }

                                "nodes" -> {
                                    respondNodeList(IoTSession(sessionID, IoTSessionType.INFO))
                                }

                                else -> respondToSession(IoTSession(sessionID, IoTSessionType.INFO), "Unknown option to get: $request")
                            }
                        }

                        else -> respondToSession(IoTSession(sessionID, IoTSessionType.OTHER), "Invalid request: $request")
                    }
                }

                // We have to flush otherwise, out data may never be sent.
                // TODO: fixing the shutdown would prevent the need for this
                _inSocket.out!!.flush()
            } catch (e: EOFException) {
                println("End of file: should be closed!")
                return  // Cleanup is done after function returns
            } catch (ex: SocketException) {
                println("Interrupted, server must be closing.")
                return  // Cleanup is done after function returns
            } catch (ex: IOException) {
                ex.printStackTrace()
                return  // Cleanup is done after function returns
            } catch (ex: JSONException) {
                ex.printStackTrace()
                println("Something is wrong with the json format.")
                return
            }
        }
    }

    private fun respondToSession(session: IoTSession, sectionList: IoTPacketSectionList) {
        try {
            _inSocket.out!!.writeBytes("ZIoT")
            _inSocket.out.writeInt(session.sessionID)
            _inSocket.out.write(sectionList.networkResponse)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun respondToSession(session: IoTSession, str: String) {
        println("Responding: $str")
        val sectionList = IoTPacketSectionList()
        sectionList.add("ZIoT")
        sectionList.add(session.sessionID)
        sectionList.add(str)
        try {
            _inSocket.out.write(sectionList.networkResponse)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun respondActionList(session: IoTSession) {
        val actionList = ioTActionDAO.enabledActions
        val sectionList = IoTPacketSectionList(actionList.size * 4 + 3)
        sectionList.add("ZIoT")
        sectionList.add(session.sessionID)
        sectionList.add(actionList.size)
        for (action in actionList) {
            sectionList.add(action.uuid!!)
            sectionList.add(action.node!!.uuid!!)
            sectionList.add(action.name!!)
            //sectionList.add(action.getReturnType())
            sectionList.add(action.securityLevel!!)
            sectionList.add(action.arguments!!)
            sectionList.add(action.encrypted!!)
            sectionList.add(action.local!!)
        }
        try {
            _inSocket.out.write(sectionList.networkResponse)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun respondNodeList(session: IoTSession) {
        val ioTNodeDAO = IoTNodeDAO()
        val nodeList = ioTNodeDAO.allNodes
        val sectionList = IoTPacketSectionList(nodeList.size * 4 + 3)
        sectionList.add("ZIoT")
        sectionList.add(session.sessionID)
        sectionList.add(nodeList.size)
        for (node in nodeList) {
            sectionList.add(node.uuid!!)
            sectionList.add(node.hostname!!)
            sectionList.add(node.nodeType!!)
            sectionList.add(node.lastHeardFrom!!)
        }
        try {
            _inSocket.out.write(sectionList.networkResponse)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun cleanup() {
        try {
            if (!_inSocket.isClosed) {
                _inSocket.close()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        private val idNumber = AtomicInteger(0)
        private val threads = Collections.synchronizedList(ArrayList<Thread>(8))
        fun getThreads(): List<Thread> {
            return threads
        }
    }
}